### PR TITLE
Add CRIU to the OpenJ9 docker file

### PR DESCRIPTION
**Pull request by vsebe**
Update docker file to install CRIU in a Linux x86-64 OpenJ9 CentOS 7 or
Ubuntu 16+ docker container when --criu option is provided.

Signed-off-by: Violeta Sebe <vsebe@ca.ibm.com>

### CODE REVIEW

**Review by tajila — APPROVED** [review]
One minor comment

File: buildenv/docker/mkdocker.sh [hunk 1]
```diff
@@ -188,6 +193,26 @@ validate_options() {
...
   esac
 
+  # If CRIU is requested, validate the architecture.
+  if [ $criu != no ] ; then
+    case "$arch" in
+      x86_64)
+        if [ $dist = centos ] ; then
+          # overwrite CRIU version
+          criu=3.12
+          if [ $version != 7 ] ; then
```
[the remark this answers is not in the document]
Comment (keithc-ca) [reply buildenv/docker/mkdocker.sh:203]: Yes, I was going to suggest the condition should be `$version = 6`.

[the remark this answers is not in the document]
Comment (vsebe) [reply buildenv/docker/mkdocker.sh]: Updated as suggested, thank you!

File: buildenv/docker/mkdocker.sh [hunk 2]
```diff
@@ -329,6 +354,27 @@ fi
...
+  echo "    protobuf-c \\"
+  echo "    protobuf-c-devel \\"
+  echo "    protobuf-compiler \\"
+  echo "    protobuf-devel \\"
+  echo "    protobuf-python \\"
+  echo "    protobuf-static \\"
+  echo "    python-ipaddress \\"
+  echo "    python3-future \\"
+  echo "    xmlto \\"
+fi
```
Comment (keithc-ca) [inline buildenv/docker/mkdocker.sh:377]: Are these needed only to compile CRIU? If so, the generated dockerfile should be restructured so these extra packages are not in the final image.

File: buildenv/docker/mkdocker.sh [hunk 3]
```diff
@@ -502,6 +548,31 @@ fi
...
+  echo "    protobuf-c-compiler \\"
+  echo "    protobuf-compiler \\"
+  echo "    python-ipaddress \\"
+  echo "    python-protobuf \\"
+if [ $version != 16.04 ] ; then
+  echo "    python3-distutils \\"
+fi
+  echo "    python3-future \\"
+  echo "    xmlto \\"
+fi
```
Comment (keithc-ca) [inline buildenv/docker/mkdocker.sh:575]: Except possibly for 20.04, just installing the `criu` package should be sufficient for Ubuntu.

File: buildenv/docker/mkdocker.sh [hunk 4]
```diff
@@ -644,6 +715,24 @@ done
...
+  echo ""
+  echo "# Install CRIU and set capabilities."
+  echo "RUN cd /tmp \\"
+  echo " && $wget_O v${criu}.tar.gz https://github.com/checkpoint-restore/criu/archive/refs/tags/v${criu}.tar.gz \\"
+  echo " && tar -xzf v${criu}.tar.gz \\"
+  echo " && cd criu-${criu} \\"
+  echo " && make \\"
+  echo " && make install \\"
+  echo " && cd .. \\"
+  echo " && rm -rf v${criu}.tar.gz criu-${criu} \\"
```
Comment (keithc-ca) [inline buildenv/docker/mkdocker.sh:728]: Please don't include the version in the local archive name (just use `criu.tar.gz`, following the pattern for other packages).

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: buildenv/docker/mkdocker.sh (hunk 3 above)
Comment (vsebe) [reply buildenv/docker/mkdocker.sh]: Is not the required version. This provides the same version on all platforms.

**Comment by AdamBrousseau** [comment]
@keithc-ca anything else needed here?

**Comment by keithc-ca** [comment]
> @keithc-ca anything else needed here?

Yes. I've asked that things that are only needed to compile the criu package not be retained in the final image. I don't think we're there yet.

File: buildenv/docker/mkdocker.sh (hunk 3 above)
Comment (vsebe) [reply buildenv/docker/mkdocker.sh]: These packages are required to compile and run libcriu.

Comment (keithc-ca) [reply buildenv/docker/mkdocker.sh]: On my Ubuntu 18.04 system, I have the criu package installed, but none of
```
libbsd-dev
libcap-dev
libgnutls28-dev
libnet1-dev
libnl-3-dev
libprotobuf-c-dev
libprotobuf-dev
protobuf-c-compiler
protobuf-compiler
python-ipaddress
```
I conclude those packages are not required to run applications that use criu.

I would much prefer that the package manager be used as opposed to building from source.

**Comment by keithc-ca** [comment]
For Ubuntu 16.04, I see these warnings building CRUI:
```
Makefile.config:45: Warn: you have no libnftables installed
Makefile.config:46: Warn: Building without nftables support
```
Should they be addressed?

**Review by keithc-ca — APPROVED** [review]

**Comment by keithc-ca** [comment]
I think the result is reasonable; criu adds about 20MB to x86 images:
```
$ docker images
REPOSITORY         TAG       IMAGE ID       CREATED             SIZE
openj9/cent7       latest    fa855776a09a   About an hour ago   5.19GB
openj9/cent7criu   latest    2250ddd16c51   4 hours ago         5.21GB
openj9/ub16        latest    ca75c5c8ec47   23 hours ago        3.88GB
openj9/ub16criu    latest    21cd8897b1cd   24 hours ago        3.9GB
```

**Pull request merged**